spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
-    LET new == [ height |-> height + 1, lastCommit |-> lastCommit,
-                 VP |-> last.NextVP, NextVP |-> NextVP ] IN
-    /\ IsCorrectPower(NextVP) \* the correct validators have >2/3 of power
-    /\ blockchain' = Append(blockchain, new)
-    /\ height' = height + 1                   
-
-Init ==
-  /\ height = 0
-  /\ minTrustedHeight \in 1..ULTIMATE_HEIGHT
-  (* pick a genesis block of all nodes where next correct validators have >2/3 of power *)
-  /\ \E NextV \in SUBSET AllNodes \ {}:
-       \E NextVP \in [NextV -> Powers]:
-      /\ IsCorrectPower(NextVP)
-      /\  LET VP == [n \in Corr |-> 1] 
-              genesis == [ height |-> 0, lastCommit |-> {},
-                           VP |-> VP, NextVP |-> NextVP]
+(* the events that can be issued by the lite client *)    
+OutEvents ==
+        (* request the header for a given height from the peer full node *)
+    [type: "requestHeader", height: BC!Heights]
+        \union
+        (* finish the check with a verdict *)
+    [type: "finish", verdict: BOOLEAN]  
+        \union
+    {NoEvent}      
+
+(* Produce a request event for the top element of the requestStack *)
+RequestHeaderForTopRequest(stack) ==
+    IF stack = <<>>
+    THEN NoEvent
+    ELSE  LET top == Head(stack)
+              heightToRequest ==
+                IF top.isLeft
+                THEN top.endHeight      \* the pivot is on the right
+                ELSE top.startHeight    \* the pivot is on the left
           IN
-          blockchain = <<genesis>>
+          [type |-> "requestHeader", height |-> heightToRequest]
+
+(* When starting the light client *)
+OnStart ==
+    /\ state = "init"
+    /\ inEvent.type = "start"
+    /\ state' = "working"
+        (* the block at trusted height is obtained by the user *)
+        \* TODO: the English spec does not make this explicit
+    /\ storedHeaders' = { << blockchain[TRUSTED_HEIGHT],
+                             DOMAIN blockchain[TRUSTED_HEIGHT].VP >> }
+        (* The only request on the stack ("left", h1, h2) *)
+    /\ LET initStack == << [isLeft |-> TRUE,
+                           startHeight |-> TRUSTED_HEIGHT,
+                           endHeight |-> inEvent.heightToVerify] >>
+       IN           
+        /\ requestStack' = initStack
+        /\ outEvent' = RequestHeaderForTopRequest(initStack)
+
+(* Check whether we can trust the signedHdr based on trustedHdr
+   following the trusting period method.
+   This operator is similar to CheckSupport in the English spec.
+   
+   The parameters have the following meanings:
+   - heightToTrust is the height of the trusted header
+   - heightToVerify is the height of the header to be verified
+   - trustedHdr is the trusted header (not a signed header)
+   - signedHdr is the signed header to verify (including commits)
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
+    IF minTrustedHeight > heightToTrust \* outside of the trusting period
```

Comment:
What if we are just at the border line (in terms of trusted period) and this condition evaluates to false, but then during the execution of this function the trusted header leaves trusted period? Is there a way that we add a check that might catch this (potential) bug?

Reply:
Why is it a bug? We should understand expectations on time constraints.